docs/dev/guide/custom-aa-tutorial.md

```diff
@@ -0,0 +1,637 @@
+    function deployAccount(
+        bytes32 salt,
+        address owner1,
+        address owner2
+    ) external returns (address) {
+        return DEPLOYER_SYSTEM_CONTRACT.create2AA(salt, aaBytecodeHash, 0, abi.encode(owner1, owner2));
+    }
+}
+```
+
+Note, that on zkSync, the deployment is not done via bytecode, but via bytecode hash. The bytecode itself is passed to the operator via `factoryDeps` field. Note, that the `_aaBytecodeHash` must be formed in a special way:
+
+- Firstly, it is hashed with sha256
+- Then, the first two bytes are replaced with the length of the bytecode in 32-byte words.
+
+You don't need to worry about it, since our SDK provides a built-in method to do it, that will be explained below.
+
+## Deploying the factory
+
+To deploy a factory, we need to create a deployment script. Create the `deploy` folder and create one file there: `deploy-factory.ts`. Put the following deployment script there:
+
+```ts
+import { utils, Wallet } from "zksync-web3";
+import * as ethers from "ethers";
+import { HardhatRuntimeEnvironment } from "hardhat/types";
+import { Deployer } from "@matterlabs/hardhat-zksync-deploy";
+
+// An example of a deploy script that will deploy and call a simple contract.
+export default async function (hre: HardhatRuntimeEnvironment) {
+    const wallet = new Wallet("<PRIVATE-KEY>");
+    const deployer = new Deployer(hre, wallet);
+    const factoryArtifact = await deployer.loadArtifact("AAFactory");
+    const aaArtifact = await deployer.loadArtifact("TwoUserMultisig");
+
+    // Deposit some funds to L2 in order to be able to perform L2 transactions.
+    // You can remove the depositing step if you know that the `wallet` has enough
+    // funds on zkSync.
+    const depositAmount = ethers.utils.parseEther("0.001");
+    const depositHandle = await deployer.zkWallet.deposit({
+        to: deployer.zkWallet.address,
+        token: utils.ETH_ADDRESS,
+        amount: depositAmount,
+    });
+    await depositHandle.wait();
+
+    // Getting the bytecodeHash of the AA
+    const bytecodeHash = utils.hashBytecode(aaArtifact.bytecode);
+
+    const factory = await deployer.deploy(factoryArtifact, [bytecodeHash], undefined, [
+        // Since the factory requires the code for the AA to be available,
+        // we should pass it here as well
+        aaArtifact.bytecode
+    ]);
+
+    console.log(`AA factory address: ${factory.address}`);
+}
+```
+
+In order to deploy the factory, you should run compile the contracts and run the script:
+
+```
+yarn hardhat compile
+yarn hardhat deploy-zksync --script deploy-factory.ts
+```
+
+The output should be roughly the following:
+
+```
+AA factory address: 0x9db333Cb68Fb6D317E3E415269a5b9bE7c72627Ds
+```
+
+Where the address will be different for each run.
+
+## Working with accounts
+
+### Deploying account
+
+Now, let's deploy an AA and initiate a new transaction with it. In this section we already assume that you have an EOA wallet with enough funds on it.
```

Comment:
```suggestion
Now, let's deploy an account and initiate a new transaction with it. In this section we assume that you already have an EOA account with enough funds on it.

Reply:
On L2. From the user's perspective it is indistinguishable from an EOA, so I would avoid saying "EOA-like" to avoid confusion

Reply:
well, I generally agree, and for me this is understandable. However, I do think that we generate more confusion here, as on the one hand we talk about abstractions of accounts -- there are no more EOA _or_ contracts -- but on the other we keep using this terminology.

Maybe "default EOA logic"? or "default account logic"

Reply:
No, I would say that we have three citizens here:

- EOA
- Contract
- AA

The EOA *might* get different behavior, since we want to support `tx.origin` and if we want to preserve the semantics of `tx.origin == msg.sender`, then we'll need to keep it different

Reply:
Words like "default EOA logic" or "default account logic" imply that the user had to deploy such a wallet, but they don't need to

Reply:
Doesn't the first 2 just specific instationations of AA?

Reply:
mmm... default means default -- nothing particular to do. I think

Reply:
> Doesn't the first 2 just specific instationations of AA?

We can put it this way :) 

> mmm... default means default -- nothing particular to do. I think

I just don't like that describing it takes 3 words, sounds too complicated for me

Reply:
your call